src/socket/icmpv4.rs

```diff
@@ -0,0 +1,378 @@
+use core::cmp;
+use managed::Managed;
+
+use {Error, Result};
+use phy::DeviceCapabilities;
+use socket::{Socket, SocketHandle};
+use storage::{Resettable, RingBuffer};
+use wire::{IpProtocol, Ipv4Address, Ipv4Repr, Icmpv4Repr, Icmpv4Packet};
+
+/// A buffered ICMPv4 packet.
+#[derive(Debug)]
+pub struct PacketBuffer<'a> {
+    addr:    Ipv4Address,
+    size:    usize,
+    payload: Managed<'a, [u8]>
+}
+
+impl<'a> PacketBuffer<'a> {
+    /// Create a buffered packet.
+    pub fn new<T>(payload: T) -> PacketBuffer<'a>
+            where T: Into<Managed<'a, [u8]>> {
+        PacketBuffer {
+            addr:     Ipv4Address::default(),
+            size:     0,
+            payload:  payload.into()
+        }
+    }
+
+    fn as_ref<'b>(&'b self) -> &'b [u8] {
+        &self.payload[..self.size]
+    }
+
+    fn as_mut<'b>(&'b mut self) -> &'b mut [u8] {
+        &mut self.payload[..self.size]
+    }
+
+    fn resize<'b>(&'b mut self, size: usize) -> Result<&'b mut Self> {
+        if self.payload.len() >= size {
+            self.size = size;
+            Ok(self)
+        } else {
+            Err(Error::Truncated)
+        }
+    }
+}
+
+impl<'a> Resettable for PacketBuffer<'a> {
+    fn reset(&mut self) {
+        self.size = 0;
+    }
+}
+
+/// An ICMPv4 packet ring buffer.
+pub type SocketBuffer<'a, 'b: 'a> = RingBuffer<'a, PacketBuffer<'b>>;
+
+/// An ICMPv4 socket
+///
+/// An ICMPv4 socket is bound to a specific  destination ip and
+/// owns transmit and receive packet buffers.
+#[derive(Debug)]
+pub struct Icmpv4Socket<'a, 'b: 'a> {
+    handle:    SocketHandle,
+    rx_buffer: SocketBuffer<'a, 'b>,
+    tx_buffer: SocketBuffer<'a, 'b>,
+    /// The endpoint this socket is communicating with
+    dst_addr:  Ipv4Address,
+    /// The time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ttl:       Option<u8>
```

Comment:
Since this socket works with raw packets the TTL is already in there.

Reply:
We still use `IpRepr::emit`, so we still need to populate the ttl of the repr [here](https://github.com/dlrobertson/smoltcp/blob/icmpv4_socket/src/socket/icmpv4.rs#L209-L224) right? It is a "raw packet" but I made the assumption that the raw data passed to `send` would not include the IP header. That may be an error in my implementation.

Reply:
Ah, true.